# Homework 2: Data Wrangling with Pandas
## Max Chelminski

In [1]:
import pandas as pd
import numpy as np

In [2]:
#Loading data
stats = pd.read_csv('../../Data/Seasons_Stats_NBA.csv')
stats

,Year,Player,Pos,Age,Tm,G,GS,MP,PER,TS%,...,FT%,ORB,DRB,TRB,AST,STL,BLK,TOV,PF,PTS
0,1950.0,Curly Armstrong,G-F,31.0,FTW,63.0,NaN,NaN,NaN,0.368,...,0.705,NaN,NaN,NaN,176.0,NaN,NaN,NaN,217.0,458.0
1,1950.0,Cliff Barker,SG,29.0,INO,49.0,NaN,NaN,NaN,0.435,...,0.708,NaN,NaN,NaN,109.0,NaN,NaN,NaN,99.0,279.0
2,1950.0,Leo Barnhorst,SF,25.0,CHS,67.0,NaN,NaN,NaN,0.394,...,0.698,NaN,NaN,NaN,140.0,NaN,NaN,NaN,192.0,438.0
3,1950.0,Ed Bartels,F,24.0,TOT,15.0,NaN,NaN,NaN,0.312,...,0.559,NaN,NaN,NaN,20.0,NaN,NaN,NaN,29.0,63.0
4,1950.0,Ed Bartels,F,24.0,DNN,13.0,NaN,NaN,NaN,0.308,...,0.548,NaN,NaN,NaN,20.0,NaN,NaN,NaN,27.0,59.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24686,2017.0,Cody Zeller,PF,24.0,CHO,62.0,58.0,1725.0,16.7,0.604,...,0.679,135.0,270.0,405.0,99.0,62.0,58.0,65.0,189.0,639.0
24687,2017.0,Tyler Zeller,C,27.0,BOS,51.0,5.0,525.0,13.0,0.508,...,0.564,43.0,81.0,124.0,42.0,7.0,21.0,20.0,61.0,178.0
24688,2017.0,Stephen Zimmerman,C,20.0,ORL,19.0,0.0,108.0,7.3,0.346,...,0.600,11.0,24.0,35.0,4.0,2.0,5.0,3.0,17.0,23.0
24689,2017.0,Paul Zipser,SF,22.0,CHI,44.0,18.0,843.0,6.9,0.503,...,0.775,15.0,110.0,125.0,36.0,15.0,16.0,40.0,78.0,240.0


### Q1: Which NBA player scored the most points in a single season? What seasons/year was this?

In [3]:
stats[stats['PTS'] == stats['PTS'].max()]

,Year,Player,Pos,Age,Tm,G,GS,MP,PER,TS%,...,FT%,ORB,DRB,TRB,AST,STL,BLK,TOV,PF,PTS
1706,1962.0,Wilt Chamberlain*,C,25.0,PHW,80.0,NaN,3882.0,31.7,0.536,...,0.613,NaN,NaN,2052.0,192.0,NaN,NaN,NaN,123.0,4029.0


Wilt Chamberlain scored the most points in a single season, with 4029 points in 1962.

### Q2: Which NBA player scored the most points in 1991?

In [4]:
stats[stats['Year'] == 1991].sort_values(by='PTS', ascending=False).head(1)

,Year,Player,Pos,Age,Tm,G,GS,MP,PER,TS%,...,FT%,ORB,DRB,TRB,AST,STL,BLK,TOV,PF,PTS
10212,1991.0,Michael Jordan*,SG,27.0,CHI,82.0,82.0,3034.0,31.6,0.605,...,0.851,118.0,374.0,492.0,453.0,223.0,83.0,202.0,229.0,2580.0


Michael Jordan scored the most points in 1991 with 2580 points.

### Q3: Which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data? If there are multiple players who have tied for the best percentage, then sort on the column with player names and submit your answer as a `pd.DataFrame object`.

In [5]:
stats_2000plus = stats[stats['Year'] >= 2000]
answer_q3 = stats_2000plus[stats_2000plus['FT%'] == stats_2000plus['FT%'].max()].sort_values(by='Player')
answer_q3[['Player', 'FT%']]

,Player,FT%
17990,Aaron Miles,1.0
16181,Adam Harrington,1.0
22787,Adonis Thomas,1.0
22788,Adonis Thomas,1.0
18961,Adrian Griffin,1.0
...,...,...
16906,Wesley Person,1.0
18271,Will Blalock,1.0
17572,Yuta Tabuse,1.0
17167,Zarko Cabarkapa,1.0


There are 313 players who had a free throw percentage of 100%, as shown by the above DataFrame.

### Q4: Among those with at least 50 attempts ('FTA'), which player had the best free throw percentage ("FT%") from the year 2000 to the most recent year in the data?

In [6]:
stats_q4 = stats_2000plus[stats_2000plus['FTA'] >= 50].sort_values(by='FT%', ascending=False).head(1)
stats_q4[['Player', 'FTA', 'FT%']]

,Player,FTA,FT%
15565,Terrell Brandon,84.0,0.988


Terrell Brandon had the best free throw percentage of players from the year 2000 onward with 98.8% over 84 free throws.

### Q5: Rename the variable "Pos" to "Position".

In [7]:
stats = stats.rename(columns={'Pos': 'Position'})
stats['Position'].head(3)

0    G-F
1     SG
2     SF
Name: Position, dtype: object

### Q6: Use this variable to create two variables that are called "first_position" and "second_position". Hint: If a player plays two positions, then the "-" is used as a delimiter between their first and second positions. In these cases, you will have to split the position variable into two.

In [8]:
stats[['first_position', 'second_position']] = stats['Position'].str.split(pat='-', expand=True)
stats[['first_position', 'second_position']].head(1)

,first_position,second_position
0,G,F


In [9]:
positions = {'PF': 'Power Forward', 'SF': 'Strong Forward', 'SG': 'Shooting Guard', 'PG': 'Point Guard', 'G': 'Guard', 'C': 'Center', 'F': 'Forward'}
stats[['first_position', 'second_position']] = stats[['first_position', 'second_position']].replace(positions)
stats[['first_position', 'second_position']].head(5)

,first_position,second_position
0,Guard,Forward
1,Shooting Guard,None
2,Strong Forward,None
3,Forward,None
4,Forward,None


### Q7: Combine these two variables back into a single variable called "position_united". Make sure to separate the two positions by "-", as the original data field did. If there is no second position, then there will only be a single position. Again, this is similar to how the original variable was coded. Present a frequency table of "position_united".

In [10]:
stats['position_united'] = np.where(stats['second_position'].isnull(), stats['first_position'], stats['first_position'] + '-' + stats['second_position'])
stats[['position_united', 'first_position', 'second_position']].head(3)

,position_united,first_position,second_position
0,Guard-Forward,Guard,Forward
1,Shooting Guard,Shooting Guard,None
2,Strong Forward,Strong Forward,None


In [11]:
stats['position_united'].value_counts()

position_united
Power Forward                    4966
Shooting Guard                   4811
Center                           4759
Strong Forward                   4699
Point Guard                      4648
Guard                             139
Forward                           123
Forward-Center                     81
Guard-Forward                      79
Forward-Guard                      60
Center-Forward                     35
Center-Power Forward               33
Strong Forward-Shooting Guard      30
Point Guard-Shooting Guard         28
Power Forward-Center               27
Shooting Guard-Point Guard         27
Shooting Guard-Strong Forward      26
Power Forward-Strong Forward       25
Strong Forward-Power Forward       21
Shooting Guard-Power Forward        3
Center-Strong Forward               2
Strong Forward-Point Guard          1
Point Guard-Strong Forward          1
Name: count, dtype: int64

### Q8: Create two new datasets.
* a new dataset from the original dataset that includes all data except the age variable (be sure to give this dataset a new name).
* a new dataset from the original dataset that includes the year, the player name, and age.
* add a new column to both datasets called mergeid that includes a sequence of numbers beginning with a 1 in the first row of the data and ending with the total number of rows in the last row of the data

In [12]:
stats_q8_1 = stats.drop(columns='Age')
stats_q8_2 = stats.loc[:,['Year', 'Player', 'Age']]

stats_q8_1['mergeid'] = range(1, len(stats_q8_1) + 1)
stats_q8_2['mergeid'] = range(1, len(stats_q8_2) + 1)

print(stats_q8_1['mergeid'])
print(stats_q8_2['mergeid'])

0            1
1            2
2            3
3            4
4            5
         ...  
24686    24687
24687    24688
24688    24689
24689    24690
24690    24691
Name: mergeid, Length: 24691, dtype: int64
0            1
1            2
2            3
3            4
4            5
         ...  
24686    24687
24687    24688
24688    24689
24689    24690
24690    24691
Name: mergeid, Length: 24691, dtype: int64


### Q9: Join the two datasets from question (8) together to recreate the original dataset plus the new merge id.

In [13]:
stats_q9 = pd.merge(stats_q8_1, stats_q8_2, on=['Year', 'Player', 'mergeid'])
stats_q9

,Year,Player,Position,Tm,G,GS,MP,PER,TS%,3PAr,...,STL,BLK,TOV,PF,PTS,first_position,second_position,position_united,mergeid,Age
0,1950.0,Curly Armstrong,G-F,FTW,63.0,NaN,NaN,NaN,0.368,NaN,...,NaN,NaN,NaN,217.0,458.0,Guard,Forward,Guard-Forward,1,31.0
1,1950.0,Cliff Barker,SG,INO,49.0,NaN,NaN,NaN,0.435,NaN,...,NaN,NaN,NaN,99.0,279.0,Shooting Guard,None,Shooting Guard,2,29.0
2,1950.0,Leo Barnhorst,SF,CHS,67.0,NaN,NaN,NaN,0.394,NaN,...,NaN,NaN,NaN,192.0,438.0,Strong Forward,None,Strong Forward,3,25.0
3,1950.0,Ed Bartels,F,TOT,15.0,NaN,NaN,NaN,0.312,NaN,...,NaN,NaN,NaN,29.0,63.0,Forward,None,Forward,4,24.0
4,1950.0,Ed Bartels,F,DNN,13.0,NaN,NaN,NaN,0.308,NaN,...,NaN,NaN,NaN,27.0,59.0,Forward,None,Forward,5,24.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
24686,2017.0,Cody Zeller,PF,CHO,62.0,58.0,1725.0,16.7,0.604,0.002,...,62.0,58.0,65.0,189.0,639.0,Power Forward,None,Power Forward,24687,24.0
24687,2017.0,Tyler Zeller,C,BOS,51.0,5.0,525.0,13.0,0.508,0.006,...,7.0,21.0,20.0,61.0,178.0,Center,None,Center,24688,27.0
24688,2017.0,Stephen Zimmerman,C,ORL,19.0,0.0,108.0,7.3,0.346,0.000,...,2.0,5.0,3.0,17.0,23.0,Center,None,Center,24689,20.0
24689,2017.0,Paul Zipser,SF,CHI,44.0,18.0,843.0,6.9,0.503,0.448,...,15.0,16.0,40.0,78.0,240.0,Strong Forward,None,Strong Forward,24690,22.0


### Q10: Subset the original dataset to the years 1995-1997. Group the data by team name and year and then summarize the average number of points per team. Arrange from most to least points. (Hint: Your answer should have a shape of 88 x 3, since we are looking at the average points scored over the 1995-1997 seasons. Note that 2 of these columns could be indexes, in which case Pandas will tell you your data is 88 rows x 1 columns. Finally, also note that the data for VAN and TOR are missing in 1995 as these teams did not exist in that year. So your answer could also be 90 x 3, if you include missing/empty data in these rows).

In [14]:
stats_q10 = stats[(stats['Year'] >= 1995) & (stats['Year'] <= 1997)]
stats_q10.groupby(['Tm', 'Year'])['PTS'].mean().reset_index(name='PTS_per_Year').sort_values('PTS_per_Year', ascending=False)

,Tm,Year,PTS_per_Year
73,SEA,1996.0,659.384615
72,SEA,1995.0,646.785714
47,MIN,1997.0,606.307692
54,ORL,1995.0,606.066667
60,PHO,1995.0,604.866667
...,...,...,...
20,DEN,1997.0,348.695652
50,NJN,1997.0,346.695652
58,PHI,1996.0,322.750000
79,TOT,1997.0,322.380952


### Q11: Similar to Q10, start by subsetting the original dataset to the years 1995-1997. Then create a wide dataset that keeps each year in its own row, but spreads team names to multiple columns with each column delineating points per team in each of the 1995-1997 seasons (Hint: Your final answer should have a shape of 3 x 30).

In [15]:
stats_q11 = stats_q10.copy()
stats_q11 = stats_q11.pivot_table(index='Year', columns='Tm', values='PTS')
stats_q11

Tm,ATL,BOS,CHH,CHI,CLE,DAL,DEN,DET,GSW,HOU,...,PHO,POR,SAC,SAS,SEA,TOR,TOT,UTA,VAN,WSB
Year,,,,,,,,,,,,,,,,,,,,,
1995.0,440.055556,495.764706,515.562500,520.312500,494.466667,604.428571,488.764706,503.312500,481.500000,499.470588,...,604.866667,528.187500,537.066667,546.375000,646.785714,NaN,488.681818,581.733333,NaN,515.125000
1996.0,447.722222,471.944444,401.476190,575.000000,498.200000,560.600000,471.352941,558.714286,490.235294,466.888889,...,503.058824,543.000000,510.187500,565.133333,659.384615,420.736842,419.767857,600.285714,409.0000,400.380952
1997.0,457.294118,458.222222,450.444444,563.866667,551.769231,275.222222,348.695652,482.687500,510.687500,458.222222,...,366.565217,427.052632,494.250000,370.900000,517.125000,372.809524,322.380952,563.600000,457.0625,543.133333


### Q12: Now return the data from #11 back into a long (tidy) format by moving Team (TM), Points and Year into their own columns (Hint: Your data shape should be 90 x 3).

In [16]:
stats_q12 = stats_q11.copy()
stats_q12.reset_index().melt(id_vars='Year', value_name='PTS')

,Year,Tm,PTS
0,1995.0,ATL,440.055556
1,1996.0,ATL,447.722222
2,1997.0,ATL,457.294118
3,1995.0,BOS,495.764706
4,1996.0,BOS,471.944444
...,...,...,...
85,1996.0,VAN,409.000000
86,1997.0,VAN,457.062500
87,1995.0,WSB,515.125000
88,1996.0,WSB,400.380952
